### PR TITLE
[FIXED JENKINS-33614] Add link to script approval when a rejection occurs

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-33614](https://issues.jenkins-ci.org/browse/JENKINS-33614)

I don't *love* this, so feel free to critique/veto.

cc @reviewbybees

### CODE REVIEW

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/Messages.properties [hunk 1]
```diff
@@ -1 +1,2 @@
 Snippetizer.this_step_should_not_normally_be_used_in=This step should not normally be used in your script. Consult the inline help for details.
+SandboxContinuable.ScriptApprovalLink=Administrators can click here to approve or reject this signature.
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/Messages.properties]: Add the final newline.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/SandboxContinuable.java [hunk 2]
```diff
@@ -36,6 +42,13 @@ public Outcome call() {
                     RejectedAccessException x = findRejectedAccessException(outcome.getAbnormal());
                     if (x != null) {
                         ScriptApproval.get().accessRejected(x, ApprovalContext.create());
+                        try {
+                            e.getOwner().getListener().getLogger().println(x.getMessage() + ". " +
+                                    HyperlinkNote.encodeTo("/" + ScriptApproval.get().getUrlName(),
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/SandboxContinuable.java]: More work, but it is possible to create a custom note which would check whether the viewer has `RUN_SCRIPTS` or not, and only conditionally hyperlink it.

Comment (abayer) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/SandboxContinuable.java]: ooooo, now I'm gonna go figure out how to do that!

File: src/main/java/org/jenkinsci/plugins/workflow/cps/SandboxContinuable.java [hunk 3]
```diff
@@ -72,5 +79,52 @@ public Outcome call() {
...
+                String url = "/" + ScriptApproval.get().getUrlName();
+
+                StaplerRequest req = Stapler.getCurrentRequest();
+
+                if (req!=null) {
+                    // if we are serving HTTP request, we want to use app relative URL
+                    url = req.getContextPath()+url;
+                } else {
+                    // otherwise presumably this is rendered for e-mails and other non-HTTP stuff
+                    url = Jenkins.getInstance().getRootUrl()+url.substring(1);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/SandboxContinuable.java]: A little roundabout to `substring` out a char we just appended, but OK.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/SandboxContinuable.java [hunk 4]
```diff
@@ -72,5 +79,52 @@ public Outcome call() {
         }
     }
 
+    public static final class ScriptApprovalNote extends ConsoleNote {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/SandboxContinuable.java]: Can be `private` I think.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java [hunk 5]
```diff
@@ -174,6 +180,15 @@ public void superCallsSandboxed() throws Exception {
 
     @Test
     public void sandboxInvokerUsed() throws Exception {
+        jenkins.jenkins.setSecurityRealm(jenkins.createDummySecurityRealm());
+        GlobalMatrixAuthorizationStrategy gmas = new GlobalMatrixAuthorizationStrategy();
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java]: Use `MockAuthorizationStrategy`. Much easier.

**Pull request merged**